src/renderer_gl.cpp

```diff
@@ -1092,180 +1092,116 @@ namespace bgfx { namespace gl
 		GL_CHECK(glUseProgram(program) );
 	}
 
+	struct f4 { float val[4]; bool operator ==(const f4 &rhs) { const uint64_t *a = (const uint64_t *)this; const uint64_t *b = (const uint64_t *)&rhs; return a[0] == b[0] && a[1] == b[1]; }};
+	struct f3x3 { float val[9]; bool operator ==(const f3x3 &rhs) { const uint64_t *a = (const uint64_t *)this; const uint64_t *b = (const uint64_t *)&rhs; return a[0] == b[0] && a[1] == b[1] && a[2] == b[2] && a[3] == b[3] && ((const uint32_t*)a)[8] == ((const uint32_t*)b)[8]; }};
+	struct f4x4 { float val[16]; bool operator ==(const f4x4 &rhs) { const uint64_t *a = (const uint64_t *)this; const uint64_t *b = (const uint64_t *)&rhs; return a[0] == b[0] && a[1] == b[1] && a[2] == b[2] && a[3] == b[3] && a[4] == b[4] && a[5] == b[5] && a[6] == b[6] && a[7] == b[7]; }};
+
+	// Inserts the new value into the uniform cache, and returns true
+	// if the old value was different than the new one.
 	template<typename T>
-	class TinyHashMap
+	static bool updateUniformCache(uint32_t loc, const T &value)
 	{
-		struct KeyVal
-		{
-			uint64_t key;
-			T value;
-		};
-		KeyVal *table;
-		uint32_t capacityMask;
-		uint32_t size;
-
-	public:
-		TinyHashMap()
-		:table(0), capacityMask(0), size(0)
-		{
-			Reserve();
-		}
-
-		~TinyHashMap()
-		{
-			delete[] table;
-		}
+		// Uniform state cache for various types.
+		static tinystl::unordered_map<uint64_t, T> uniformCacheMap;
```

Comment:
`static tinystl::unordered_map<uint64_t, T> uniformCacheMap;`

Do not make it static, make it part of GL renderer context so that it can be deallocated properly.

Also use `stl::unordered_map`.

Reply:
Also make some `#if BGFX_GL_CONFIG_UNIFORM_CACHE` so that this code path can be enabled/disabled.